File: src/System/Mesos/Raw/ContainerInfo.hs
{-# LANGUAGE OverloadedStrings #-}
module System.Mesos.Raw.ContainerInfo where

import           System.Mesos.Internal
import           System.Mesos.Raw.Volume

type ContainerInfoPtr = Ptr ContainerInfo

foreign import ccall unsafe "ext/types.h toContainerInfo" c_toContainerInfo
  :: CInt -- ^ type
  -> Ptr CChar -- ^ image
  -> CInt -- ^ image length
  -> Ptr VolumePtr -- ^ volumes
  -> CInt -- ^ volumes length
  -> IO ContainerInfoPtr

foreign import ccall unsafe "ext/types.h fromContainerInfo" c_fromContainerInfo
  :: ContainerInfoPtr
  -> Ptr CInt -- ^ type
  -> Ptr (Ptr CChar) -- ^ image
  -> Ptr CInt -- ^ image length
  -> Ptr (Ptr VolumePtr)
  -> Ptr CInt -- ^ volumes length
  -> IO ()

foreign import ccall unsafe "ext/types.h destroyContainerInfo" c_destroyContainerInfo
  :: ContainerInfoPtr -> IO ()

instance CPPValue ContainerInfo where
  marshal ci = do
    (ip, il) <- cstring $ dockerImage $ containerInfoContainerType ci
    (vp, vl) <- mapM marshal (containerInfoVolumes ci) >>= arrayLen
    liftIO $ c_toContainerInfo 1 ip (fromIntegral il) vp (fromIntegral vl)

  unmarshal p = do
    tyP <- alloc
    (iP, ilP) <- arrayPair
    vPP <- alloc
    vlP <- alloc
    liftIO $ c_fromContainerInfo p tyP iP ilP vPP vlP
    ty <- peek tyP
    im <- peekMaybeCString (iP, ilP)
    vP <- peek vPP
    vps <- peekArray (vP, vlP)
    vs <- mapM unmarshal vps
    if ty == 1
       then return $ ContainerInfo (Docker $ maybe "" id im) vs
       else return $ ContainerInfo (Unknown $ fromIntegral ty)  vs

  destroy = c_destroyContainerInfo

File: ext/types/container_info.cpp
#include <iostream>
#include "types.h"

using namespace mesos;

ContainerInfoPtr toContainerInfo (int type,
				  char* image,
				  int imageLen,
				  VolumePtr* volumes,
				  int volumesLen)
{
  ContainerInfoPtr p = new ContainerInfo;
  p->set_type((ContainerInfo_Type) type);
  if (type == ContainerInfo_Type_DOCKER)
    {
      ContainerInfo_DockerInfo d;
      d.set_image(image, imageLen);
      *p->mutable_docker() = d;
    }
  
  for (int i = 0; i < volumesLen; ++i)
    *p->add_volumes() = *volumes[i];

  return p;
}

void fromContainerInfo (ContainerInfoPtr info,
			int* type,
			char** image,
			int* imageLen,
			VolumePtr** volumes,
			int* volumesLen)
{
  *type = info->type();

  if (info->type() == 1) // DOCKER
    {
      *image = (char*) info->mutable_docker()->mutable_image()->data();
      *imageLen = info->mutable_docker()->mutable_image()->size();
    }
  
  *volumes = (VolumePtr*) info->mutable_volumes()->data();
  *volumesLen = info->mutable_volumes()->size();
}

void destroyContainerInfo (ContainerInfoPtr containerInfo)
{
  delete containerInfo;
}
